sycl/include/CL/sycl/bit_cast.hpp

```diff
@@ -0,0 +1,67 @@
+//==---------------- bit_cast.hpp - SYCL bit_cast --------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <type_traits>
+
+#if __cpp_lib_bit_cast
+#include <bit>
+#endif
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+
+// forward decl
+namespace detail {
+inline void memcpy(void *Dst, const void *Src, std::size_t Size);
+}
+
+// sycl::bit_cast ( no longer sycl::detail::bit_cast )
+template <typename To, typename From>
+#if __cpp_lib_bit_cast || __has_builtin(__builtin_bit_cast)
+constexpr
+#endif
+    To
+    bit_cast(const From &from) noexcept {
+  static_assert(sizeof(To) == sizeof(From),
+                "Sizes of To and From must be equal");
+  static_assert(std::is_trivially_copyable<From>::value,
+                "From must be trivially copyable");
+  static_assert(std::is_trivially_copyable<To>::value,
+                "To must be trivially copyable");
+#if __cpp_lib_bit_cast
+  return std::bit_cast<To>(from);
+#else // __cpp_lib_bit_cast
+
+#if __has_builtin(__builtin_bit_cast)
+  return __builtin_bit_cast(To, from);
+#else  // __has_builtin(__builtin_bit_cast)
+  static_assert(std::is_trivially_default_constructible<To>::value,
+                "To must be trivially default constructible");
```

Comment:
Just wondering: shouldn't this condition be checked regardless compiler built-ins support?

Reply:
I did not write this code, I just moved it out of helpers.hpp.  However, I think the answer is that that std::bit_cast only requires `is_trivially_copyable`.  But, if you are implementing it yourself with `memcpy` that implementation will also require `is_trivially_default_constructible`.  The code in the "Notes" section here suggests this: https://en.cppreference.com/w/cpp/numeric/bit_cast .

Reply:
Actually it says that additional requirement is `is_trivially_constructible`, not `is_trivially_default_constructible`.
And our implementation requires `is_trivially_constructible` even if `std::bit_cast` is supported. According to the cppreference it's not necessary...

Reply:
Sorry, I mixed up `is_trivially_copyable` and `is_trivially_constructible` in my previous comment. Please, ignore it.